src/Microsoft.Health.Fhir.CosmosDb.UnitTests/Features/Storage/Queues/CosmosQueueClientTests.cs

```diff
@@ -0,0 +1,134 @@
+    {
+        // Arrange
+        ICosmosQuery<JobGroupWrapper> cosmosQuery = Substitute.For<ICosmosQuery<JobGroupWrapper>>();
+        _cosmosQueryFactory.Create<JobGroupWrapper>(Arg.Any<Container>(), Arg.Any<CosmosQueryContext>())
+            .ReturnsForAnyArgs(cosmosQuery);
+
+        int callCount = 0;
+        cosmosQuery.ExecuteNextAsync(Arg.Any<CancellationToken>()).ReturnsForAnyArgs(_ =>
+        {
+            if (callCount++ == 0)
+            {
+                throw new TestCosmosException(statusCode);
+            }
+
+            return Task.FromResult(Substitute.For<FeedResponse<JobGroupWrapper>>());
+        });
+
+        // Act
+        await _cosmosQueueClient.DequeueAsync(0, "testworker", 10, CancellationToken.None);
+
+        // Assert
+        Assert.Equal(2, callCount);
+        await cosmosQuery.ReceivedWithAnyArgs(2).ExecuteNextAsync(Arg.Any<CancellationToken>());
+    }
+
+    [Theory]
+    [InlineData(typeof(CosmosException))]
+    [InlineData(typeof(RequestRateExceededException))]
+    public async Task GivenADequeueJobOperation_WhenExceptionWithRetryAfterIsProvided_PolicyRespectsRetryAfter(Type exceptionType)
+    {
+        // Arrange
+        ICosmosQuery<JobGroupWrapper> cosmosQuery = Substitute.For<ICosmosQuery<JobGroupWrapper>>();
+        _cosmosQueryFactory.Create<JobGroupWrapper>(Arg.Any<Container>(), Arg.Any<CosmosQueryContext>())
+            .ReturnsForAnyArgs(cosmosQuery);
+        var retryAfter = TimeSpan.FromSeconds(2);
+        int callCount = 0;
+
+        cosmosQuery.ExecuteNextAsync(Arg.Any<CancellationToken>()).ReturnsForAnyArgs(_ =>
+        {
+            if (callCount++ == 0)
+            {
+                throw exceptionType == typeof(CosmosException)
+                    ? new TestCosmosException(HttpStatusCode.TooManyRequests, retryAfter)
+                    : new RequestRateExceededException(retryAfter);
+            }
+
+            return Task.FromResult(Substitute.For<FeedResponse<JobGroupWrapper>>());
+        });
+
+        var stopwatch = Stopwatch.StartNew();
+
+        // Act
+        await _cosmosQueueClient.DequeueAsync(0, "testworker", 10, CancellationToken.None);
+
+        stopwatch.Stop();
+
+        // Assert
+        Assert.Equal(2, callCount);
+        await cosmosQuery.ReceivedWithAnyArgs(2).ExecuteNextAsync(Arg.Any<CancellationToken>());
+
+        // Allowing small imprecision due to timer resolution
+        var actualElapsedSeconds = stopwatch.Elapsed.TotalSeconds;
+        Assert.True(
+            Math.Abs(actualElapsedSeconds - retryAfter.TotalSeconds) <= 0.5,
+            $"Expected retry after {retryAfter.TotalSeconds} seconds, but actual elapsed time was {actualElapsedSeconds} seconds.");
+    }
+
+    public class TestCosmosException(HttpStatusCode statusCode, TimeSpan? retryAfter = null) : CosmosException("Test exception message", statusCode, 0, "test-activity-id", 0.0)
```

Comment:
Not sure if Health Paas will be able to compile this part of the code. We are still using .NET 6 for retro compatibility, right? I remember I did something like that two years ago, and it broke Health Paas.

Reply:
Great catch - I'm not sure if Roslyn will handle this for us - but not worth it. I'm always tempted to use new features. 😄